jlm/rvsdg/FunctionType.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2017 Nico Reißmann <nico.reissmann@gmail.com>
+ * Copyright 2025 Helge Bahmann <hcb@chaoticmind.net>
+ * See COPYING for terms of redistribution.
+ */
+
+#include <sstream>
+
+#include <jlm/rvsdg/FunctionType.hpp>
+#include <jlm/util/Hash.hpp>
+
+namespace jlm::rvsdg
+{
+
+FunctionType::~FunctionType() noexcept = default;
+
+FunctionType::FunctionType(
+    std::vector<std::shared_ptr<const jlm::rvsdg::Type>> argumentTypes,
+    std::vector<std::shared_ptr<const jlm::rvsdg::Type>> resultTypes)
+    : jlm::rvsdg::ValueType(),
+      ArgumentTypes_(std::move(argumentTypes)),
+      ResultTypes_(std::move(resultTypes))
+{}
+
+std::string
+FunctionType::debug_string() const
+{
+  std::stringstream ss;
+  ss << "fun (";
+  bool first = true;
+  for (const auto & argtype : ArgumentTypes_)
+  {
+    if (first)
+    {
+      ss << ", ";
+      first = false;
+    }
+    ss << argtype->debug_string();
+  }
+
+  ss << ") -> (";
+  for (const auto & restype : ResultTypes_)
```

Comment:
I think you forgot to set first back to true here.

Reply:
Not only that, but part of the logic was backwards. Fixed it now.